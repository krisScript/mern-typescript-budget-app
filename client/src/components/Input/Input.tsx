import React, {
  useState,
  useMemo,
  FunctionComponent,
  SetStateAction,
  Dispatch,
} from 'react';

interface InputProps {
  name: string;
  placeholder: string;
  value: string;
  type: string;
  validationErrorParams: string[];
  setValueHook: Dispatch<SetStateAction<string>>;
}
const Input: FunctionComponent<InputProps> = ({
  name,
  placeholder,
  value,
  type,
  setValueHook,
  validationErrorParams,
}): JSX.Element => {
  const error = useMemo(() => validationErrorParams.includes(name), [
    validationErrorParams,
    name,
  ]);
  return (
    <div className="field">
      <label className="label">{placeholder}</label>
      <div className="control">
        <input
          className={`input${error ? ' is-danger' : ''}`}
          onChange={e => setValueHook(e.target.value)}
          value={value}
          type={type}
          placeholder={placeholder}
          name={name}
          required
        />
      </div>
    </div>
  );
};
export default Input;
