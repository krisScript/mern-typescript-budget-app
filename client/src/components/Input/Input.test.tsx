/* eslint-disable @typescript-eslint/explicit-function-return-type */
import React from 'react';

import { render, fireEvent } from 'react-testing-library';
import Input from './Input';

describe('<Input />', () => {
  const setValueHook = jest.fn();
  const value = 'John Doe';
  const type = 'text';
  const placeholder = 'Test Placeholder';
  const name = 'username';
  const userInputValue = 'newUsername';
  let validationErrorParams: string[] = [];
  const { container, getByPlaceholderText, rerender } = render(
    <Input
      setValueHook={setValueHook}
      value={value}
      type={type}
      placeholder={placeholder}
      validationErrorParams={validationErrorParams}
      name={name}
    />,
  );
  it('should render', async () => {
    const inputElement = getByPlaceholderText(placeholder) as HTMLInputElement;
    expect(inputElement).toBeTruthy;
    expect(inputElement.value).toEqual(value);
    expect(inputElement.className).toEqual('input');
    expect(inputElement.type).toEqual(type);
    expect(inputElement.name).toEqual(name);
  });
  describe('onChange Event', () => {
    beforeEach(() => {
      const inputElement = getByPlaceholderText(
        placeholder,
      ) as HTMLInputElement;
      fireEvent.change(inputElement, { target: { value: userInputValue } });
    });
    it('setValueHook should be called', () => {
      expect(setValueHook).toBeCalled();
      expect(setValueHook).toBeCalledTimes(1);
    });
  });
});
