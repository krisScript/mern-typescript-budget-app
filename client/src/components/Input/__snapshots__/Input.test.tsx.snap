// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Input /> snapshot 1`] = `
<div>
  <div
    class="field"
  >
    <label
      class="label"
    >
      username
    </label>
    <div
      class="control"
    >
      <input
        class="input"
        name="username"
        placeholder="Test Placeholder"
        type="text"
        value="John Doe"
      />
    </div>
  </div>
</div>
`;
